From c53de2c8e359493d3a4a80b037f5c1cd9cbdf1da Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty
 <45927311+Etaash-mathamsetty@users.noreply.github.com>
Date: Thu, 23 Jul 2026 21:56:32 -0400
Subject: [PATCH 225/268] win32u: Improve implementation of AMD
 KMTQAITYPE_UMDRIVERPRIVATE.

Now should support AGS and amdxcffx64
---
 dlls/win32u/d3dkmt.c | 57 ++++++++++++++++++++++++++++++++++++--------
 1 file changed, 47 insertions(+), 10 deletions(-)

diff --git a/dlls/win32u/d3dkmt.c b/dlls/win32u/d3dkmt.c
index 9c257c6c77d..3e69cd3dc68 100644
--- a/dlls/win32u/d3dkmt.c
+++ b/dlls/win32u/d3dkmt.c
@@ -680,6 +680,31 @@ NTSTATUS WINAPI NtGdiDdDDIDestroyDevice( const D3DKMT_DESTROYDEVICE *desc )
     return STATUS_SUCCESS;
 }
 
+typedef struct
+{
+    /* 0x960 size */
+    UINT32 reserved1[2]; /* 0 */
+    UINT32 numInfo; /* 8 */
+    USHORT vendorID; /* c */
+    BYTE reserved2[6]; /* 0xE */
+    USHORT deviceID; /* x14 */
+    BYTE reserved3[0x94a];
+    /* 0x3a00 size */
+    struct
+    {
+        BYTE reserved1[0x894]; /* 0x0 */
+        UINT32 pciInfo; /* U32 @ 0x894: bus number + function number + deviceNumber */
+        BYTE reserved2[0x16c]; /* 0x898 */
+        UINT32 adapterFamily;  /* U32 @ 0xa04: adapter family */
+        UINT32 adapterRevision; /* U32 @ 0xa08: which GPU/APU it is in that family */
+        BYTE reserved3[0x238]; /* 0xa0c */
+        UINT32 unkFeature; /* U32 @ 0xc44: LSB needs to be set to 1 for RDNA */
+        UINT32 unkROP1; /* U32 @ 0xc48: numROPs = unkROP1 * unkROP2 * 4 */
+        UINT32 unkROP2; /* U32 @ 0xc4c */
+        BYTE reserved4[0x2DB0]; /* If reporting GCN architecture, there is some fields in this reserved block */
+    } info[1];
+} AMDLargeDriverPrivate;
+
 /******************************************************************************
  *           NtGdiDdDDIQueryAdapterInfo    (win32u.@)
  */
@@ -798,25 +823,37 @@ NTSTATUS WINAPI NtGdiDdDDIQueryAdapterInfo( D3DKMT_QUERYADAPTERINFO *desc )
         if (properties2.properties.deviceType == VK_PHYSICAL_DEVICE_TYPE_INTEGRATED_GPU)
             rdna2 = FALSE;
 
+        /* AGS has a bug where it reads uninitialized memory with the small private data */
         if (properties2.properties.vendorID == 0x1002 && desc->PrivateDriverDataSize == 0x260)
+            return STATUS_BUFFER_TOO_SMALL;
+
+        if (properties2.properties.vendorID == 0x1002 && desc->PrivateDriverDataSize == 0x4360)
         {
-            int *data = desc->pPrivateDriverData;
+            AMDLargeDriverPrivate *data = desc->pPrivateDriverData;
+
+            data->vendorID = properties2.properties.vendorID;
+            data->deviceID = properties2.properties.deviceID;
+            data->numInfo = 1;
+            /* we set this to 0 in ADL */
+            data->info[0].pciInfo = 0;
+            /* RDNA 1 */
+            data->info[0].adapterFamily = 0x8f;
+            data->info[0].adapterRevision = 0x2;
+            data->info[0].unkFeature = 1;
+            data->info[0].unkROP1 = 4;
+            data->info[0].unkROP2 = 4;
+
             if (fp8_support && wmma)
             {
                 /* Navi4x 9070xt */
-                data[0xc] = 0x98; /* APU/GPU Family */
-                data[0xd] = 0x51; /* Revision/which GPU it is in that family */
+                data->info[0].adapterFamily = 0x98;
+                data->info[0].adapterRevision = 0x51;
             }
             else if (rdna2 || ((e = getenv("FSR4_UPGRADE")) && *e == '1'))
             {
                 /* Navi31 */
-                data[0xc] = 0x91; /* APU/GPU Family */
-                data[0xd] = 0x3; /* Revision/which GPU it is in that family */
-            }
-            else
-            {
-                WARN("Not recommended to use FSR-I8, Use FSR4_UPGRADE=1 for FSR4-I8!\n");
-                return STATUS_NOT_IMPLEMENTED;
+                data->info[0].adapterFamily = 0x91;
+                data->info[0].adapterRevision = 0x3;
             }
 
             return STATUS_SUCCESS;
-- 
2.55.0

